numpy/lib/function_base.py

```diff
@@ -514,7 +514,10 @@ def average(a, axis=None, weights=None, returned=False):
         avg = a.mean(axis)
         scl = avg.dtype.type(a.size/avg.size)
     else:
-        a = a + 0.0
+        # Cast bool, unsigned int, and int to float64 by default
```

Comment:
Thinking somemore, this is probably a good use cars for einsum. It can do the necessary sum product without any temporaries. The only possible drawback is that we would then miss out on the pair wise summation that np.add does these days. Seems like a good trade off to me. The fastest implementation for the 1d case is probably via vdot, too FWIW.